### PR TITLE
Home Screen: Show a snackbar hint on survey change

### PR DESCRIPTION
**Pull request by scolsen**
Displays a snackbar hint when a new survey is selected. The hint text varies based on the following survey properties:

- If there is at least one job lacking the predefined data collection strategy.
- If there are zero LOIs associated with the jobs.

Otherwise, we display some default text.

Screenshot 2024-02-28 at 4 02 43 PM

Screenshot 2024-02-28 at 4 03 18 PM

@gino-m

### CODE REVIEW

**Comment by scolsen** [comment]
Fixes #2274

File: ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt [hunk 1]
```diff
@@ -128,6 +130,29 @@ class HomeScreenMapContainerFragment : AbstractMapContainerFragment() {
...
+                  it.strategy != Job.DataCollectionStrategy.UNKNOWN
+              }
+            val readonly = lois.isEmpty()
+            val msg =
+              when {
+                suggest -> resources.getString(R.string.suggest_data_collection_hint)
+                readonly -> resources.getString(R.string.read_only_data_collection_hint)
+                else -> resources.getString(R.string.predefined_data_collection_hint)
+              }
+            Snackbar.make(binding.root, msg, Snackbar.LENGTH_LONG).show()
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt:152]: Nit: We have a util for these, `EphermeralPopups`. Maybe we can move this line there?

Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt:152]: Per designs the Snackbar should stick around until dismissed. If not this in this PR, please add a TODO(bug) to add that.

File: ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt [hunk 2]
```diff
@@ -128,6 +130,29 @@ class HomeScreenMapContainerFragment : AbstractMapContainerFragment() {
     super.onViewCreated(view, savedInstanceState)
     setupMenuFab()
     setupBottomLoiCards()
+    lifecycleScope.launch {
+      mapContainerViewModel.activeSurvey
+        .combine(mapContainerViewModel.lois) { x, y -> Pair(x, y) }
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt:135]: Is this the same as identity function?

File: ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt [hunk 3]
```diff
@@ -128,6 +130,29 @@ class HomeScreenMapContainerFragment : AbstractMapContainerFragment() {
...
+        .collect {
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt:136]: Let's refactor this callback into its own fun.

File: ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt [hunk 4]
```diff
@@ -128,6 +130,29 @@ class HomeScreenMapContainerFragment : AbstractMapContainerFragment() {
...
+          val survey = it.first
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt:137]: I thing you can more idiomatically write `.collect{ (survey, lois) -> ..` here

File: ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt [hunk 5]
```diff
@@ -128,6 +130,29 @@ class HomeScreenMapContainerFragment : AbstractMapContainerFragment() {
...
+          val lois = it.second
+          survey?.let {
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt:139]: ..and perhaps add `filterNotNull` on survey above.

Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt:139]: ..and perhaps add `filterNotNull` on survey above.

Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt:139]: ..and perhaps add `filterNotNull` on survey above.

File: ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt [hunk 6]
```diff
@@ -128,6 +130,29 @@ class HomeScreenMapContainerFragment : AbstractMapContainerFragment() {
...
+            val suggest =
+              it.jobs.any {
+                it.strategy != Job.DataCollectionStrategy.PREDEFINED &&
+                  it.strategy != Job.DataCollectionStrategy.UNKNOWN
+              }
+            val readonly = lois.isEmpty()
+            val msg =
+              when {
+                suggest -> resources.getString(R.string.suggest_data_collection_hint)
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt:148]: I find the indirection in these vars and conditions hard to grok and keep in my head all at once. Instead, perhaps we can write:

```kt
EphemeralPopups.dismissableSnackbar(binding.root, zoomHintId(survey, lois))
...
fun zoomHintId(survey, lois) {
val msgId = when {
  survey.jobs.any { it.canDataCollectorsAddLois() } -> R.string.ad_hoc_data_collection zoom_hint
  lois.isEmpty() -> R.string.read_only_data_collection_zoom_hint
  else -> R.string.predefined_data_collection_zoom_hint
}

}
```

**Review by gino-m — COMMENTED** [review]
Thank you! A few suggestions, otherwise LG!

**Review by gino-m — CHANGES_REQUESTED** [review]
Thank you! A few suggestions, otherwise LG!

File: ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt (hunk 2 above)
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt:135]: Won't this trigger on every LOI change?

File: ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt (hunk 1 above)
Comment (scolsen) [reply ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt:152]: ahhh, sgtm, done

Comment (scolsen) [reply ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt:152]: I somehow missed that, done!

File: ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt (hunk 4 above)
Comment (scolsen) [reply ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt:137]: yep, good callout

File: ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt (hunk 5 above)
Comment (scolsen) [reply ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt:139]: done

File: ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt (hunk 2 above)
Comment (scolsen) [reply ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt:135]: Ahhhh yes I believe so. Switched it up to take only the `first()` batch.

File: ground/src/main/java/com/google/android/ground/ui/common/EphemeralPopups.kt [hunk 7]
```diff
@@ -16,25 +16,74 @@
...
 import androidx.annotation.StringRes
 import com.google.android.ground.R
+import com.google.android.material.snackbar.Snackbar
 import javax.inject.Inject
 import javax.inject.Singleton
 
 /** Displays short-lived messages such as toasts that are shown over other UI elements. */
 @Singleton
-class EphemeralPopups @Inject constructor(private val context: Application) {
+data class EphemeralPopups @Inject constructor(private val context: Application) {
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/ui/common/EphemeralPopups.kt:29]: I think `data class` here isn't really true to the spirit of what data classes are intended for.

File: ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt [hunk 8]
```diff
@@ -130,28 +128,26 @@ class HomeScreenMapContainerFragment : AbstractMapContainerFragment() {
...
+  }
+
+  /**
+   * Displays a popup hint informing users how to begin collecting data, based on the properties of
+   * the active survey.
+   *
+   * This method should only be called after view creation.
+   */
+  private suspend fun showDataCollectionHint() {
+    mapContainerViewModel.surveyUpdateFlow.collect {
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt:141]: We should probably check `mapContainerViewModel != null` here and log a warning if this is called before view creation.

File: ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerViewModel.kt [hunk 9]
```diff
@@ -81,13 +82,17 @@ internal constructor(
...
 
-  val lois: Flow<Set<LocationOfInterest>> =
-    activeSurvey.flatMapLatest {
-      if (it == null) {
-        flow { setOf<LocationOfInterest>() }
-      } else {
-        loiRepository.getLocationsOfInterests(it)
-      }
+  /** Captures essential, high-level derived properties for a given survey. */
+  data class SurveyProperties(val suggestLoiPermitted: Boolean, val readOnly: Boolean)
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerViewModel.kt:86]: s`/suggest/add`

File: ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt [hunk 10]
```diff
@@ -130,28 +128,26 @@ class HomeScreenMapContainerFragment : AbstractMapContainerFragment() {
...
+      val messageId =
+        when {
+          it.suggestLoiPermitted -> R.string.suggest_data_collection_hint
```
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt:144]: s/suggestLoi/addLoi/ throughout

**Review by gino-m — CHANGES_REQUESTED** [review]
A few minor comments, otherwise LG!

File: ground/src/main/java/com/google/android/ground/ui/common/EphemeralPopups.kt (hunk 7 above)
Comment (scolsen) [reply ground/src/main/java/com/google/android/ground/ui/common/EphemeralPopups.kt:29]: I agree, but detekt was forcing me to do it :(

Comment (scolsen) [reply ground/src/main/java/com/google/android/ground/ui/common/EphemeralPopups.kt:29]: I'll see if/how I can surppress it here.

File: ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt (hunk 8 above)
Comment (scolsen) [reply ground/src/main/java/com/google/android/ground/ui/home/mapcontainer/HomeScreenMapContainerFragment.kt:141]: sgtm, I was hoping android had an annotation for this (e.g. like their `WorkerThread` annotation), but alas.

**Comment by scolsen** [comment]
@gino-m should be ready for another look!

**Review by gino-m — APPROVED** [review]
Thanks for applying all the changes! LGTM!

**Review by gino-m — CHANGES_REQUESTED** [review]
Looks like Compile files.. did I merge something incorrectly
```
file:///workspace/ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/photo/PhotoTaskFragment.kt:191:14 Unresolved reference: showError
```

**Comment by scolsen** [comment]
Ah yeah, looks like a call to EphemeralPopups just needs to change.

File: ground/src/main/java/com/google/android/ground/ui/common/EphemeralPopups.kt [hunk 11]
```diff
@@ -16,25 +16,75 @@
...
-  private fun showLong(message: String) = Toast.makeText(context, message, Toast.LENGTH_LONG).show()
+    private fun showToast(@StringRes messageId: Int, duration: PopupDuration) {
+      val dur =
+        if (duration == PopupDuration.SHORT) {
+          Toast.LENGTH_SHORT
+        } else {
+          Toast.LENGTH_LONG
+        }
+      Toast.makeText(context, messageId, dur).show()
+    }
```
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/ui/common/EphemeralPopups.kt:56]: We can get rid of this helper method by reusing the other one. Something like:

`showToast(context.getString(messageId), duration)`

File: ground/src/main/java/com/google/android/ground/ui/common/EphemeralPopups.kt [hunk 12]
```diff
@@ -16,25 +16,75 @@
...
+
+    private fun showToast(message: String, duration: PopupDuration) {
+      val dur =
+        if (duration == PopupDuration.SHORT) {
+          Toast.LENGTH_SHORT
+        } else {
+          Toast.LENGTH_LONG
+        }
```
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/ui/common/EphemeralPopups.kt:64]: We're reusing the PopupDuration here but haven't handled the `INDEFINITE` separately. Can we do something about it?

Comment (scolsen) [reply ground/src/main/java/com/google/android/ground/ui/common/EphemeralPopups.kt:64]: Added a comment as to why we don't handle it for now

**Review by shobhitagarwal1612 — APPROVED** [review]

**Review by gino-m — APPROVED** [review]
Nice work!! 🚀

**Pull request merged**